### PR TITLE
Text input patterns and simple components

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR introduces two patterns: a basic text `input`, as well as a pattern that pairs a text `input` with an icon-only button (`text-input` and `text-input-with-button`, respectively). It also proposes two corresponding (logic-less)[1] components, `TextInput` and `TextInputWithButton` that wrap those patterns.

These patterns are based on existing styling in the client application, and will be used in the VitalSource content picker. See https://github.com/hypothesis/lms/issues/2862

This is the first time this package has proposed providing logic-less components, and I've done my best to make them breathtakingly basic. The `TextInput` component forwards all valid `HTMLInputElement` attributes except `type` (which it sets itself), and `TextInputWithButton` simply wraps its children in an appropriately-classed element.

I'm opening this PR in draft because:

* I'm still looking for thoughts about what to do about testing here. Ostensibly we could test for the `TextInput` applying the `is-error` class correctly. I can't really see much to do testing-wise beyond that.
* The logic-less components here are a new pattern for us, and bear some consideration.

_Edit_: moving out of draft as tests have been added, and some review has happened here.

Here is a screen shot of the pattern library page for the design patterns:

image

And a screen shot of the pattern library page for the components that wrap those patterns (`Checkbox` already existed):

image

You can see here how the correspondence of a design pattern with a component that wraps it continues to feel like an intuitive way to expose stuff.

[1] Technically `TextInput` takes an optional boolean `error` prop and will apply a CSS class when it's `true`, but that's cutting it fine on "logic".

~~This PR is current failing on test coverage, which is no kind of surprise.~~

### CODE REVIEW

**Comment by lyzadanger** [comment]
One thing that occurs to me as we consider these simple wrapping components is that perhaps they should all support taking additional arbitrary classNames so we don't end up pinning app developers into a corner. Material UI components, e.g., take a `classes` prop.

File: styles/variables/_colors.scss [hunk 1]
```diff
@@ -15,16 +15,19 @@ $grey-1: #f2f2f2;
 $grey-2: #ececec;
 $grey-3: #dbdbdb;
 $grey-4: #a6a6a6;
-$grey-5: #9c9c9c;
-$grey-6: #737373;
-$grey-7: #595959;
+$grey-5: #9c9c9c; // Fails all WCAG contrast tests; do not use for text on white
```
Comment (LMS007) [inline styles/variables/_colors.scss:18]: helpful comments

**Review by LMS007 — COMMENTED** [review]
Awesome Lyza. I like this interface for using a button with an input. It's easy to use and understand. As far as testing, I  don't think it needs to be very robust since it does not do too much. I think testing that it can just render and passes the error prop to class is sufficient as you mentioned.

A few comments...

I wonder how the focus outline will look with the button next to the input. It might be okay, but I'm not 100% sure.  Having said that,  I think it's time to fix the keyboard focus outline in the pattern lib so buttons get a focus outline when tabbing around. 

The red class when applied to the input means that when you focus on the input, you don't really get a visual affordance other then the background color. It might be nice to have a slightly darker outline on focus in that specific case? I'm just thinking out loud.

I see we're going with an `inset` box-shadow which is a change from the input and textarea in our sidebar right now. I'm okay with this, but it will mean we should probably change textarea too along with this change when this eventually gets used in the sidebar which I'm assuming it will at some point? Maybe we can discuss more tomorrow?

Other then that, I think this can more or less land as is :)

**Comment by lyzadanger** [comment]
@LMS007 

> The red class when applied to the input means that when you focus on the input, you don't really get a visual affordance other then the background color. 

Fabulous observation, really good point. This pattern is intended to be used in conjunction with error text/information, but for colorblind users, the field itself would look identical in focused and error states. Let me think on that.  Any ideas?

**Comment by lyzadanger** [comment]
> I see we're going with an inset box-shadow which is a change from the input and textarea in our sidebar right now. I'm okay with this, but it will mean we should probably change textarea too along with this change when this eventually gets used in the sidebar which I'm assuming it will at some point? Maybe we can discuss more tomorrow?

I went inset because I don't like this focus situation in the client, currently:

image

I think we can punt on figuring out `textarea` just yet, though I've provided a "stub" with the `input` mixin. We'll make sure that whatever we introduce to the client is reasonably internally consistent.

**Comment by lyzadanger** [comment]
@LMS007 

> I wonder how the focus outline will look with the button next to the input. 

Can you expand a little?

File: src/components/TextInput.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * `TextInput` sets the `type` attribute on the `input`, but any other valid
+ * attribute should be forwarded.
+ * @typedef {Omit<import('preact').JSX.HTMLAttributes<HTMLInputElement>, 'type'>} HTMLInputElementProps
+ * @typedef {TextInputBaseProps & HTMLInputElementProps} TextInputProps
+ */
+
+/**
+ * @typedef TextInputWithButtonProps
+ * @prop {Children} children
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/components/TextInput.js:20]: This component doesn't take `error`, though. :)

File: src/components/TextInput.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * Wrap a textual `<input>` element with some styles and provide error UI
+ *
+ * @param {TextInputProps} props
+ */
+export function TextInput({ error, ...restProps }) {
+  return (
+    <input
+      type="text"
+      className={classnames('Hyp-TextInput', { 'is-error': error })}
+      {...restProps}
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/components/TextInput.js:33]: That's the idea! We'll see how this pattern works out in practice...

**Comment by lyzadanger** [comment]
Tests added, and wouldn't you know it, one actually raised an issue with props ordering in `TextInput`. So they have some value!

**Comment by lyzadanger** [comment]
@LMS007 

I've added tests. Your comment about the error styling is a good one, but I don't see an obvious way to differentiate just now. Fortunately, most forms of color blindness allow the differentiation between red and blue (we'd be in more trouble with green). Not that this is a solution. I'd like to keep thinking.

File: src/components/TextInput.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (LMS007) [reply src/components/TextInput.js:20]: Yep, my bad, looking at the wrong props type!

File: src/components/test/TextInput-test.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { mount } from 'enzyme';
+
+import { TextInput, TextInputWithButton } from '../TextInput';
+
+describe('TextInput', () => {
```
Comment (LMS007) [inline src/components/test/TextInput-test.js:5]: Looks good

**Review by LMS007 — APPROVED** [review]
Okay, we'll track the use case of focus color in error state in a follow up todo or ticket and I'll take a look at getting to the bottom of why we are not seeing keyboard focus styling in the pattern lib.

GTG!

**Pull request merged**